cardano-testnet/test/cardano-testnet-test/Cardano/Testnet/Test/LedgerEvents/Gov/DRepActivity.hs

```diff
@@ -0,0 +1,351 @@
+  delegateToDRep execConfig epochStateView configurationFile socketPath sbe work "drep2-delegation"
+                 wallet2 (defaultDelegatorStakeKeyPair 2) drep2
+
+  drep3 <- registerDRep execConfig epochStateView ceo work "drep3" wallet0
+  delegateToDRep execConfig epochStateView configurationFile socketPath sbe work "drep3-delegation"
+                 wallet1 (defaultDelegatorStakeKeyPair 3) drep3
+
+  expirationDates <- checkDRepState sbe (File configurationFile) (File socketPath) execConfig
+                                    (\m -> if length m == 3 then Just $ Map.map drepExpiry m else Nothing)
+  H.note_ $ "Expiration dates for the registered DReps: " ++ show expirationDates
+
+  -- This proposal should fail because there is 2 DReps that don't vote (out of 3)
+  -- and we have the stake distributed evenly
+  let secondTargetDRepActivity = firstTargetDRepActivity + 1
+  void $ activityChangeProposalTest execConfig epochStateView configurationFile socketPath ceo gov
+                                    "failingProposal" wallet2 [(1, "yes")] secondTargetDRepActivity
+                                    minEpochsToWaitIfNotChanging (Just firstTargetDRepActivity)
+                                    maxEpochsToWaitAfterProposal
+
+  -- We now send a bunch of proposals to make sure that the 2 new DReps expire.
+  -- because DReps won't expire if there is not enough activity (opportunites to participate).
+  -- This is accounted for by the dormant epoch count
+  let numOfFillerProposals = 4 :: Int
+  sequence_
+    [activityChangeProposalTest execConfig epochStateView configurationFile socketPath ceo gov
+                                ("fillerProposalNum" ++ show proposalNum) wallet [(1, "yes")]
+                                (secondTargetDRepActivity + fromIntegral proposalNum)
+                                minEpochsToWaitIfNotChanging Nothing
+                                maxEpochsToWaitAfterProposal
+     | (proposalNum, wallet) <- zip [1..numOfFillerProposals] (cycle [wallet0, wallet1, wallet2])]
+
+  (EpochNo epochAfterTimeout) <- getCurrentEpochNo epochStateView
+  H.note_ $ "Epoch after which we are going to test timeout: " <> show epochAfterTimeout
+
+  -- Last proposal (set activity to something else again and it should pass, because of inactivity)
+  -- Because 2 out of 3 DReps were inactive, prop should pass
+  let lastTargetDRepActivity = secondTargetDRepActivity + fromIntegral numOfFillerProposals + 1
+  void $ activityChangeProposalTest execConfig epochStateView configurationFile socketPath ceo gov
+                                    "lastProposal" wallet0 [(1, "yes")] lastTargetDRepActivity
+                                    minEpochsToWaitIfChanging (Just lastTargetDRepActivity)
+                                    maxEpochsToWaitAfterProposal
+
+activityChangeProposalTest
+  :: (MonadTest m, MonadIO m, H.MonadAssertion m, MonadCatch m, Foldable t)
```

Comment:
I feel that it may be easy to make a mistake when calling this function with so many arguments. Maybe it would be better to create a separate datatype for that purpose?

Reply:
Yes, I was tempted to do a monad that holds that info and refactor all the DRep tests and functions to use it. Maybe I can do that in a PR after these are merged

Reply:
And you can probably also use a newly-introduced datatype to pass less arguments to `delegateToDRep`, for example these ones:

```hs
EpochStateView -- ^ Current epoch state view for transaction building. It can be obtained
FilePath -- ^ Path to the node configuration file as returned by 'cardanoTestnetDefault'.
FilePath -- ^ Path to the cardano-node unix socket file.
ShelleyBasedEra ConwayEra -- ^ The Shelley-based era (e.g., 'ConwayEra') in which the transaction will be constructed.
```

Reply:
Yes, if we have the base config vars in a Monad, and I think that could easily include the `era` and the `EpochStateView`, then the remaining args are going to be really few. We can still also pass a single argument that is a record so that remaining params are labeled, definitely.

Reply:
Precisely, those are the ones I meant would be nice to hide with a monad. A configuration datatype with all the config info would be an alternative too, even though it is still one extra parameter which you have to include every time. Another thing I considered was returning a record of partially applied functions that include the config, but I discarded the idea because I thought it would be pretty ugly